lib/rspec/support/object_inspector.rb

```diff
@@ -4,19 +4,39 @@ module Support
     # printing Time, DateTime, or BigDecimal
     module ObjectInspector
       # @api private
-      def self.inspect(object)
-        if Time === object
+      # rubocop:disable CyclomaticComplexity
+      # rubocop:disable MethodLength
+      def self.inspect(object, nested=false)
+        case
+        when Time === object
           format_time(object)
-        elsif defined?(DateTime) && DateTime === object
+        when defined?(DateTime) && DateTime === object
           format_date_time(object)
-        elsif defined?(BigDecimal) && BigDecimal === object
+        when defined?(BigDecimal) && BigDecimal === object
           "#{object.to_s 'F'} (#{object.inspect})"
-        elsif RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
+        when defined?(Hash) && Hash === object
+          format_hash(object, nested)
+        when defined?(Array) && Array === object
```

Comment:
`Hash` and `Array` are always defined, so there's no need for the `defined?` checks on them. More generally, if we're going to switch to a `case` statement we can leverage that directly rather than using `===`:

``` ruby
case object
when Time
  format_time(object)
when Hash
  format_hash(object, nested)
when Array
  "[#{object.map { |element| ObjectInspector.inspect(element, true) }.join(', ')}]"
else
  if defined?(DateTime) && DateTime === object
    format_date_time(object)
  elsif defined?(BigDecimal) && BigDecimal === object
    "#{object.to_s 'F'} (#{object.inspect})"
  elsif RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
    object.description
  else
    object.inspect
  end
end
```